docroot/modules/custom/foia_quarterly_data_report/src/Plugin/Action/QuarterlyReportModerationAction.php

```diff
@@ -0,0 +1,296 @@
+   *   Moderation information service.
+   * @param \Drupal\content_moderation\StateTransitionValidationInterface $validation
+   *   Validation information service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, MessengerInterface $messenger, ModerationInformationInterface $moderation_info, StateTransitionValidationInterface $validation) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->messenger = $messenger;
+    $this->moderationInfo = $moderation_info;
+    $this->validation = $validation;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('messenger'),
+      $container->get('content_moderation.moderation_information'),
+      $container->get('content_moderation.state_transition_validation')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute($entity = NULL) {
+    if (is_array($entity)) {
+      $this->executeMultiple($entity);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function executeMultiple(array $entities) {
+    $batches = [];
+    foreach ($entities as $entity) {
+      $node = $this->getLatestRevision($entity);
+      $batches[] = [
+        '\Drupal\foia_quarterly_data_report\Plugin\Action\QuarterlyReportModerationAction::batchCore',
+        [$node, 'published', count($entities)],
+      ];
+    }
+
+    // Initialize batch (to set title).
+    $batch = [
+      'title' => t('Updating Moderation State'),
+      'operations' => $batches,
+      'finished' => '\Drupal\foia_quarterly_data_report\Plugin\Action\QuarterlyReportModerationAction::batchFinished',
+      'init_message' => t('Bulk operation of moderation state change is starting.'),
+      'progress_message' => t('Processed @current out of @total. Estimated time: @estimate.'),
+      'error_message' => t('The bulk operation process has encountered an error.'),
+    ];
+    batch_set($batch);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function access($object, AccountInterface $account = NULL, $return_as_object = FALSE) {
+    $user = !isset($account) ? \Drupal::service('current_user') : $account;
+    if ($object->getEntityTypeId() !== 'node') {
+      $this->messenger->addError($this->t('Can only perform publishing on FOIA quarterly report node content.'));
+      return FALSE;
+    };
+
+    if ($object->bundle() != 'quarterly_foia_report_data') {
+      $this->messenger->addError($this->t('Can only perform publishing on FOIA quarterly report content.'));
+      return FALSE;
+    }
+
+    if ($object->moderation_state->value != 'submitted_to_oip') {
+      $this->messenger->addError($this->t('Current moderation state is not a valide state for publishing.'));
```

Comment:
As long as we're making a custom class for this, why not be specific? Eg, "The report must be 'Submitted to OIP' before it can be published."